File: GCTP_CPP/transformer.cpp
#include "transformer.h"

#include "albersconeqarea.h"
#include "lambertcc.h"
#include "mercator.h"
#include "polyconic.h"
#include "equidistantc.h"
#include "sinusoidal.h"
#include "equirectangular.h"
#include "mollweide.h"
#include "intmollweide.h"
#include "transversemercator.h"
#include "miller.h"
#include "polarstereo.h"

Transformer::Transformer()
: m_inProj(NULL), m_outProj(NULL), m_errored(false)
{
}

Transformer::~Transformer()
{
   if( m_inProj )
      delete m_inProj;

   if( m_outProj )
      delete m_outProj;
}

bool Transformer::setInput( ProjCode projectionCode )
{
   if( m_inProj )
      delete m_inProj;

   m_inProj = convertProjection( projectionCode );

   return (m_inProj != NULL );
}

bool Transformer::setInput( ProjCode projectionCode, double gctpParameters[15], ProjUnit units, ProjDatum datum )
{
   if( setInput( projectionCode ) )
   {
      m_inProj->setParams( gctpParameters );
      m_inProj->setUnits( units );
      m_inProj->setDatum( datum );
      return true;
   }
   else
      return false;
}

bool Transformer::setInput( Projection &in )
{
   return setInput( in.number(), in.params(), in.units(), in.datum() );
}

Projection* Transformer::input()
{
   return m_inProj;
}

bool Transformer::setOutput( ProjCode projectionCode )
{
   if( m_outProj )
      delete m_outProj;

   m_outProj = convertProjection( projectionCode );

   return (m_outProj != NULL );
}

bool Transformer::setOutput( ProjCode projectionCode, double gctpParameters[15], ProjUnit units, ProjDatum datum )
{
   if( setOutput( projectionCode ) )
   {
      m_outProj->setParams( gctpParameters );
      m_outProj->setUnits( units );
      m_outProj->setDatum( datum );
      return true;
   }
   else
      return false;
}

bool Transformer::setOutput( Projection &out )
{
   return setOutput( out.number(), out.params(), out.units(), out.datum() );
}

Projection* Transformer::output()
{
   return m_outProj;
}

void Transformer::transform( Coordinate* io_coord )
{
   if( m_inProj == NULL || m_outProj == NULL || io_coord->units != m_inProj->units() )
   {
      io_coord->x = 0;
      io_coord->y = 0;
      io_coord->units = (ProjUnit)0;
      m_errored = true;
      return;
   }

   Coordinate tmp;
   m_inProj->inverse( io_coord->x, io_coord->y, &(tmp.x), &(tmp.y) );
   m_outProj->forward( tmp.x, tmp.y, &(io_coord->x), &(io_coord->y) );
   io_coord->units = m_outProj->units();

   m_errored = m_inProj->errorOccured() || m_outProj->errorOccured();
}

void Transformer::transformInverse( Coordinate* io_coord )
{
   if( m_inProj == NULL || io_coord->units != m_inProj->units() )
   {
      io_coord->x = 0;
      io_coord->y = 0;
      io_coord->units = (ProjUnit)0;
      m_errored = true;
      return;
   }

   Coordinate tmp(*io_coord);
   m_inProj->inverse( tmp.x, tmp.y, &(io_coord->x), &(io_coord->y) );
   io_coord->units = DEGREE;

   m_errored = m_inProj->errorOccured();
}

void Transformer::transformForward( Coordinate* io_coord )
{
   if( m_outProj == NULL || io_coord->units != DEGREE )
   {
      io_coord->x = 0;
      io_coord->y = 0;
      io_coord->units = (ProjUnit)0;
      m_errored = true;
      return;
   }

   Coordinate tmp(*io_coord);
   m_outProj->forward( tmp.x, tmp.y, &(io_coord->x), &(io_coord->y) );
   io_coord->units = m_outProj->units();

   m_errored = m_outProj->errorOccured();
}

bool Transformer::errored()
{
   return m_errored;
}

Projection* Transformer::convertProjection( ProjCode projectionCode )
{
   Projection* proj = NULL;

   switch( projectionCode )
   {
   case ALBERS:
      proj = new AlbersConEqArea();
      break;
   case LAMCC:
      proj = new LambertCC();
      break;
   case TM:
      proj = new TransverseMercator();
      break;
   case MERCAT:
      proj = new Mercator();
      break;
   case POLYC:
      proj = new Polyconic();
      break;
   case EQUIDC:
      proj = new EquidistantC();
      break;
   case SNSOID:
      proj = new Sinusoidal();
      break;
   case EQRECT:
      proj = new Equirectangular();
      break;
   case MOLL:
      proj = new Mollweide();
      break;
   case IMOLL:
      proj = new IntMollweide();
      break;
   case MILLER:
      proj = new Miller();
      break;
   case PS:
      proj = new PolarStereo();
      break;
   }

   return proj;
}

File: JGCTP/Transformer.java
package jgctp;

public class Transformer
{
   private Projection inv_proj;

   private Projection for_proj;

   public void setInverseProjection( Projection proj )
   {
      inv_proj = proj;
   }

   public void setInverseProjection( ProjCode projCode )
   {
      inv_proj = getProjection( projCode );
   }

   public Projection inverseProjection()
   {
      return inv_proj;
   }

   public void setForwardProjection( Projection proj )
   {
      for_proj = proj;
   }

   public void setForwardProjection( ProjCode projCode )
   {
      for_proj = getProjection( projCode );
   }

   public Projection forwardProjection()
   {
      return for_proj;
   }
   
   public void transform( CoordPoint inPoint, CoordPoint outPoint )
   {
      GeoPoint gPoint = new GeoPoint();
      
      try
      {
         inv_proj.inverse(inPoint, gPoint);
         for_proj.forward(gPoint, outPoint);
      }
      catch( Exception e )
      {
         e.printStackTrace();
      }
   }
   
   public void transformInverse( CoordPoint inPoint, GeoPoint outPoint )
   {
      try
      {
         for_proj.inverse(inPoint, outPoint);
      }
      catch( Exception e )
      {
         e.printStackTrace();
      }
   }
   
   public void transformForward( GeoPoint inPoint, CoordPoint outPoint )
   {
      try
      {
         for_proj.forward(inPoint, outPoint);
      }
      catch( Exception e )
      {
         e.printStackTrace();
      }
   }

   public static Projection getProjection( ProjCode projCode )
   {
      switch (projCode.val())
      {
      case 0:
         return null;
      case 1:
         return new UTM();
      case 2:
         return null;
      case 3:
         return new AlbersConEqArea();
      case 4:
         return new LambertCC();
      case 5:
         return new Mercator();
      case 6:
         return new PolarStereo();
      case 7:
         return new Polyconic();
      case 8:
         return new EquidistantC();
      case 9:
         return new TransverseMercator();
      case 10:
         return new Stereo();
      case 11:
         return new LambertAzimuthal();
      case 12:
         return new AzEquidistant();
      case 13:
         return null;
      case 14:
         return new Orthographic();
      case 15:
         return new GenVertNSP();
      case 16:
         return new Sinusoidal();
      case 17:
         return new Equirectangular();
      case 18:
         return new Miller();
      case 19:
         return new VanDerGrinten();
      case 20:
         return new HotineObMerc();
      case 21:
         return new Robinson();
      case 22:
         return new SpaceObMerc();
      case 23:
         return null;
      case 24:
         return new GoodeH();
      case 25:
         return new Mollweide();
      case 26:
         return new IntMollweide();
      case 27:
         return new Hammer();
      case 28:
         return new WagnerIV();
      case 29:
         return new WagnerVII();
      case 30:
         return new OblatedEqArea();
      }

      return null;
   }
}
